grade.go

```diff
@@ -399,51 +415,86 @@ func main() {
 
 	for _, dir := range dirs {
 		var result util.SubmissionResult
-		results.Results[dir] = &result
-		results.Order = append(results.Order, dir)
-		result.Student = dir
+		studentInfo := util.ParseStudentOgInfo(workDir + "/" + dir + "/oginfo.json")
+
+		if studentInfo.StudentEuid != "" {
+			result.Student = studentInfo.StudentEuid
+		} else {
+			result.Student = dir
+		}
+
+		results.Results[result.Student] = &result
+		results.Order = append(results.Order, result.Student)
+
 		result.AssignmentId = assignmentId
 		// find hydratedStudent information from EUID (dirname)
-		hydratedStudent := db.GetStudentByEuid(supabase, dir)
+		hydratedStudent := db.GetStudentByEuid(supabase, result.Student)
 		// if student doesn't exist, commit to db
 		if hydratedStudent.Id == 0 {
-			hydratedStudent.Euid = dir
-			hydratedStudent.Email = fmt.Sprintf("%s@unt.edu", dir) // all students have euid@unt.edu
+			hydratedStudent.Euid = result.Student
 
-			fmt.Printf("%8s: ", dir)
+			if studentInfo.StudentEmail != "" {
+				hydratedStudent.Email = studentInfo.StudentEmail
+			} else {
+				hydratedStudent.Email = fmt.Sprintf("%s@unt.edu", result.Student) // all students have euid@unt.edu
+			}
+
+			fmt.Printf("%8s: ", result.Student)
 			fmt.Printf("%+v\n", hydratedStudent)
 
-			if !isDryRun {
+			if !isDryRun || !assignmentInfo.DryRun {
 				hydratedStudent.Save(supabase)
 			}
 		}
 
 		result.StudentId = hydratedStudent.Id
 
-		result.Feedback = make([]string, len(ogInfo.Tests))
+		result.Feedback = make([]string, len(assignmentInfo.Tests))
 
-		for i, test := range ogInfo.Tests {
+		for i, test := range assignmentInfo.Tests {
 			expected := util.GetFile(workDir + "/.spec/" + test.Expected)
-			fmt.Print("Expected Output: ", expected, "\n\n")
+
+			if test.Open {
+				fmt.Printf("Test #%d Expected Output:\n%s\n\n", i+1, expected)
+			}
+
 			var input = []string{}
 			if test.Input != "" {
 				input = parseInFile(workDir + "/.spec/" + test.Input)
 			}
 
 			gradeSubmission(&result, dir, workDir, runArgs, expected, language, input, wall, i)
 
+			var testResult string
+			if result.CompileSuccess && result.Feedback[i] == "" {
+				testResult = "PASS"
+			} else {
+				testResult = "FAIL"
```

Comment:
does golang support string unions like typescript?

```ts
type TestResult = "PASS" | "FAIL"
```
for example

could be worth looking into, not needed for this pr though

Reply:
from what i can see they dont support it. i also don't quite yet understand their value

Reply:
just makes it easier to narrow types, if there's no support it's not a big deal at all